protocols/autonat/src/v2/server/handler/dial_request.rs

```diff
@@ -0,0 +1,349 @@
+use std::{
+    io,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use either::Either;
+use futures::{
+    channel::{mpsc, oneshot},
+    AsyncRead, AsyncWrite, SinkExt, StreamExt,
+};
+use futures_bounded::FuturesSet;
+use libp2p_core::{
+    upgrade::{DeniedUpgrade, ReadyUpgrade},
+    Multiaddr,
+};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    handler::{ConnectionEvent, FullyNegotiatedInbound, ListenUpgradeError},
+    ConnectionHandler, ConnectionHandlerEvent, StreamProtocol, SubstreamProtocol,
+};
+use rand_core::RngCore;
+
+use crate::v2::{
+    generated::structs::{mod_DialResponse::ResponseStatus, DialStatus},
+    protocol::{Coder, DialDataRequest, DialRequest, DialResponse, Request, Response},
+    server::behaviour::StatusUpdate,
+    Nonce, DIAL_REQUEST_PROTOCOL,
+};
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum DialBackStatus {
+    /// Failure during dial
+    DialErr,
+    /// Failure during dial back
+    DialBackErr,
+}
+
+#[derive(Debug)]
+pub struct DialBackCommand {
+    pub(crate) addr: Multiaddr,
+    pub(crate) nonce: Nonce,
+    pub(crate) back_channel: oneshot::Sender<Result<(), DialBackStatus>>,
+}
+
+pub struct Handler<R> {
+    client_id: PeerId,
+    observed_multiaddr: Multiaddr,
+    dial_back_cmd_sender: mpsc::Sender<DialBackCommand>,
+    dial_back_cmd_receiver: mpsc::Receiver<DialBackCommand>,
+    status_update_sender: mpsc::Sender<StatusUpdate>,
+    status_update_receiver: mpsc::Receiver<StatusUpdate>,
+    inbound: FuturesSet<Result<(), io::Error>>,
+    rng: R,
+}
+
+impl<R> Handler<R>
+where
+    R: RngCore,
+{
+    pub(crate) fn new(client_id: PeerId, observed_multiaddr: Multiaddr, rng: R) -> Self {
+        let (dial_back_cmd_sender, dial_back_cmd_receiver) = mpsc::channel(10);
```

Comment:
If we are already using a channel here then we could move the `Receiver` up to the behaviour. That would remove one layer of indirection because we could directly send the `DialOpts` and wouldn't need to go through `ConnectionHandlerEvent::NotifyBehaviour`.

What do you think?

Reply:
I can change it, but it is a refactor and we don't have much time left. An upside to the channel is that the control flow is maybe a little bit cleaner. but I will do whatever you recommend

Reply:
We can leave it as is yeah.